Scala perf optimizations

isaacl opened the pull request:
Major improvement for Clock encode, some small speed up for move encoding as well.

- Remove Array.tabulate and .foreach in hot paths as they require Autoboxing.
- Adjustments based on javap source inspection and JMH benchmarks.

```text
                Java    Master   This
Huffman.Encode  8.57ms  8.68ms   8.27ms
Huffman.Decode  7.37ms  5.64ms   5.62ms
Clock.Encode    185ns   312ns    192ns
Clock.Decode    206ns   231ns    220ns
```

isaacl committed:
Scala perf optimizations

- Remove Array.tabulate and .foreach in hot paths
  as they require Autoboxing.
- Adjustments based on javap source inspection and
  JMH benchmarks.

```text
                Java    Master   This
Huffman.Encode  8.57ms  8.68ms   8.27ms
Huffman.Decode  7.37ms  5.64ms   5.62ms
Clock.Encode    185ns   312ns    192ns
Clock.Decode    206ns   231ns    220ns
```

ornicar commented on src/main/scala/clock/Encoder.scala:
```diff
@@ -6,7 +6,7 @@ import org.lichess.compression.BitOps
 object Encoder:
 
   def encode(centis: Array[Int], startTime: Int): Array[Byte] =
-    if centis.isEmpty then return Array.emptyByteArray
```
I'm all for optimizing hot paths, but that's not one. It's invoked only once per game. So I think we should use the idiomatic `isEmpty`, which btw is supposed to be inlined:

```scala
@`inline` def isEmpty: Boolean = xs.length == 0
```

isaacl replied:
Once per ply

ornicar commented on src/main/scala/BitOps.scala:
```diff
@@ -1,11 +1,18 @@
+    var i   = 0
+    while i < len do
+      writeSigned(values(i), writer)
+      i += 1
```
This is not a hot path, it runs once per game. I don't think we should be re-implementing scala's `foreach` where it can't possibly make a difference.

isaacl replied:
Once per ply -- and yes, it does make a difference in the OverallEncodingTest.testEncode bench.  (260->190ns/op)

ornicar commented:
b733ef4f0f `sbt 'benchmarks/jmh:run -i 5 -wi 3 -f1 -t1 org.lichess.compression.benchmark.*'`

```
[info] Benchmark                       Mode  Cnt     Score    Error  Units
[info] BitOpsTest.testRead             avgt    5   106.057 ±  0.401  ns/op
[info] HuffmanPgnBench.decode          avgt    5  4176.123 ± 40.581  us/op
[info] HuffmanPgnBench.encode          avgt    5  7065.805 ± 28.124  us/op
[info] LinearEstimateTest.testEncode   avgt    5    32.803 ±  0.167  ns/op
[info] LowBitTruncTest.testEncode      avgt    5     3.645 ±  0.011  ns/op
[info] OverallEncodingTest.testDecode  avgt    5   129.727 ±  1.060  ns/op
[info] OverallEncodingTest.testEncode  avgt    5   159.204 ±  3.736  ns/op
[info] VarIntEncodingTest.testDecode   avgt    5    83.967 ±  0.284  ns/op
[info] VarIntEncodingTest.testEncode   avgt    5   125.071 ±  7.903  ns/op
```

ornicar committed:
idiomatic scala in functions called once per game

ornicar commented:
ffe3e19e93bd0b1bbb2fc871d8ad10b7ba1744f5 `sbt 'benchmarks/jmh:run -i 5 -wi 3 -f1 -t1 org.lichess.compression.benchmark.OverallEncodingTest'`

```
[info] Benchmark                       Mode  Cnt    Score   Error  Units
[info] OverallEncodingTest.testDecode  avgt    5  130.638 ± 3.088  ns/op
[info] OverallEncodingTest.testEncode  avgt    5  157.204 ± 2.064  ns/op
```
as expected within statistical error

ornicar commented:
Also I'm rather confused because I just tried benchmarking master 877666f40ed9566e4b52844bf6bef402563ddbff and got this:

877666f40ed9566e4b52844bf6bef402563ddbff `sbt 'benchmarks/jmh:run -i 5 -wi 3 -f1 -t1 org.lichess.compression.benchmark.OverallEncodingTest'`

```
[info] Benchmark                       Mode  Cnt    Score   Error  Units
[info] OverallEncodingTest.testDecode  avgt    5  128.926 ± 2.996  ns/op
[info] OverallEncodingTest.testEncode  avgt    5  141.193 ± 1.757  ns/op
```
Showing similar/better perf than this PR.

Using `openjdk 21.0.9 / Linux 6.17.2`.

I'm sure this PR should be better, I just don't understand how it can't be measured on my machine.

isaacl commented:
That's odd -- I get consistently better benchmarks on my machine, most of the changes here I benchmarked individually, with the exception of a couple `foreach` array replacements that I did in one go.

I did see a separate odd quirk in `jmh`, where Huffman.encode sometimes runs with 8.5ms/run and sometimes with 5.6ms/run and  restarting the run seems to flip a coin of which one it's going to be.

isaacl committed:
Cosmetic fixes

- remove @inline annotations which do nothing
- better line spacing for BitMask field

isaacl committed:
Add back optimizations

According to my JMH benches, these do matter.
idk

ornicar commented on src/main/scala/clock/LinearEstimator.scala:
```diff
@@ -2,11 +2,11 @@ package org.lichess.compression.clock
 
 object LinearEstimator:
 
-  @inline def encode(dest: Array[Int], startTime: Int): Unit =
```
since scala3 we have an `inline` keyword that guarantees inlining https://docs.scala-lang.org/scala3/guides/macros/inline.html#inline-methods

ornicar commented:
https://monitor.lichess.ovh/d/nAXOnXrWz/game-pgn-encoder?orgId=1&from=now-7d&to=now&timezone=utc&var-instance=lila@ocean&var-field=mean&refresh=5m&viewPanel=panel-2

isaacl commented:
@ornicar will investigate. Definitely not expected.
